nix/shared.nix

```diff
@@ -156,11 +156,58 @@ let
         export CPATH=${libtorch}/include/torch/csrc/api/include
       '';
     };
+  altdev-announce = libtorch: old: with builtins; with pkgs.lib.strings; with pkgs.lib.lists;
+    let
+      echo = str: "echo \"${str}\"";
+      nl = echo "";
+      findFirstPrefix = pre: def: xs: findFirst (x: hasPrefix pre x) def xs;
+      removeStrings = strs: xs: replaceStrings strs (map (x: "") strs) xs;
+
+      # findAndReplaceLTS :: [String] -> String -- something like "lts-14.7"
+      findAndReplaceLTS = xs:
+        let pre = "resolver:";
+        in removeStrings [" " "\n" pre] (findFirstPrefix pre "resolver: lts-14.7" xs);
+    in
+      old // {
+        shellHook = old.shellHook + concatStringsSep "\n" [
+          nl
+          (echo "Suggested NixOS development uses cabal v1-*. If you plan on developing on NixOS")
+          (echo "with stack, you may still need to add the following to your stack.yaml:")
+          nl
+          (echo "  extra-lib-dirs:")
+          (echo "    - ${libtorch}/lib")
+          (echo "  extra-include-dirs:")
+          (echo "    - ${libtorch}/include")
+          (echo "    - ${libtorch}/include/torch/csrc/api/include")
+          nl
+          (echo "cabal v2-* development on NixOS may also need an updated cabal.project.local:")
+          nl
+          (echo "  package libtorch-ffi")
+          (echo "    extra-lib-dirs:     ${libtorch}/lib")
+          (echo "    extra-include-dirs: ${libtorch}/include")
+          (echo "    extra-include-dirs: ${libtorch}/include/torch/csrc/api/include")
+          # zlib.out and zlib.dev are strictly for developing with a nix-shell using stack- or cabal v2- based builds.
+          # this is a similar patch to https://github.com/commercialhaskell/stack/issues/2975
+          (echo "  package zlib")
+          (echo "    extra-lib-dirs: ${pkgs.zlib.dev}/lib")
+          (echo "    extra-lib-dirs: ${pkgs.zlib.out}/lib")
+          nl
+          (echo "as well as a freeze file from stack's resolver:")
+          # $(which curl) is used to bypass an alias to 'curl'. This is safe so long as we use gnu's which
+          (echo "$(which curl) https://www.stackage.org/${findAndReplaceLTS (splitString "\n" (readFile ../stack.yaml))}/cabal.config \\")
+          (echo "  | sed -e 's/inline-c ==.*,/inline-c ==0.9.0.0,/g' -e 's/inline-c-cpp ==.*,/inline-c-cpp ==0.4.0.0,/g' \\")
+          (echo "  > cabal.project.freeze")
```

Comment:
this doesn't seem to be properly escaped. I am seeing
```
as well as a freeze file from stack's resolver:
sed: can't read ": No such file or directory
  > cabal.project.freeze
```

Reply:
I was calling `nix-shell` from within the `hasktorch` folder:
```
$ cd ~/Projects/thirdParty/hasktorch/hasktorch
$ nix-shell --arg cudaVersion 10
```

Reply:
maybe there's a way to get a stable reference to `readFile ../stack.yaml` that doesn't depend on the current working directory

Reply:
I've fixed the escaping. With respect to relative imports, nix always treats relative paths as relative to the source file as opposed to the current working directory (even symlink'd are evaluated and references are resolved correctly). They are very stable in my experience and are a common pattern in the NixOS codebase.